tests/agent_server/test.py

```diff
@@ -0,0 +1,293 @@
+
+        while after is not None:
+            result = await client.get_resource_actions(
+                tid=environment,
+                resource_type=resource_type,
+                agent=agent,
+                attribute=key,
+                attribute_value=value,
+                limit=1,
+                first_timestamp=after,
+            )
+            assert result.code == 200
+
+            actions = result.result.get("data", [])
+            actions.reverse()
+
+            for action in actions:
+                change = action.get("change") or "nochange"
+                if change != "nochange":
+                    return {
+                        "change": change,
+                        "finished": action.get("finished"),
+                        "version": action.get("version"),
+                    }
+
+            if len(actions) == 0:
+                after = None
+            else:
+                after = actions[-1].get("started")
+
+        return None
+
+    last_change_1 = await get_last_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key1",
+    )
+
+    last_change_2 = await get_last_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key2",
+    )
+
+    last_change_3 = await get_last_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key3",
+    )
+
+    first_change_1 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key1",
+    )
+
+    first_change_2 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key2",
+    )
+
+    first_change_3 = await get_first_change(
+        resource_type="test::Resource",
+        agent="agent1",
+        key="key",
+        value="key3",
+    )
+
+    import json
+    logger.info("Last change for key1: " + json.dumps(last_change_1, indent=2))
+    logger.info("Last change for key2: " + json.dumps(last_change_2, indent=2))
+    logger.info("Last change for key3: " + json.dumps(last_change_3, indent=2))
+    logger.info("First change for key1: " + json.dumps(first_change_1, indent=2))
+    logger.info("First change for key2: " + json.dumps(first_change_2, indent=2))
+    logger.info("First change for key3: " + json.dumps(first_change_3, indent=2))
+
+
+    async def get_dependencies(id: str) -> str:
+        logger.info(id)
+        result = await client.get_resource(environment, id=id, logs=True)
+        assert result.code == 200
+
+        return result.result.get("resource", {}).get("attributes", {}).get("requires")
+
+
+    async def need_redeploy(
+        resource_type: str,
+        agent: str,
+        key: str,
+        value: str,
+    ) -> bool:
+        last_deployment = await get_last_change(resource_type, agent, key, value)
```

Comment:
you don't need a change, I think, an effective deploy is sufficient (I mean, any deploy that is not due to incremental deploy)

Reply:
Should be good too, related to this: https://github.com/inmanta/inmanta-core/pull/2762#discussion_r589337499